abstract_domains/numerical/interval.py

```diff
@@ -0,0 +1,237 @@
+        self._lower = None
+        self._upper = None
+        super().__init__()
+        self._lower = lower
+        self._upper = upper
+
+    @property
+    def interval(self):
+        return self.lower, self.upper
+
+    @interval.setter
+    def interval(self, bounds):
+        (lower, upper) = bounds
+        self.lower = lower
+        self.upper = upper
+
+    def add(self, other):
+        self.interval = (self.lower + other.lower, self.upper + other.upper)
+        return self
+
+    def sub(self, other):
+        self.interval = (self.lower - other.upper, self.upper - other.lower)
+        return self
+
+    def mult(self, other):
+        comb = [self.lower * other.lower, self.lower * other.upper, self.upper * other.lower, self.upper * other.upper]
+        self.interval = (min(comb), max(comb))
+        return self
+
+    def negate(self):
+        self.interval = (min(-self.lower, -self.upper), max(-self.lower, -self.upper))
+        return self
+
+    @property
+    def lower(self):
+        return self._lower
+
+    @lower.setter
+    def lower(self, b):
+        self._lower = b
+
+    @property
+    def upper(self):
+        return self._upper
+
+    @upper.setter
+    def upper(self, b):
+        self._upper = b
+
+    def __repr__(self):
+        return f"[{self.lower},{self.upper}]"
+
+
+class IntervalLattice(Interval, BottomMixin):
+    def __repr__(self):
+        if self.is_bottom():
+            return "BOTTOM"
+        else:
+            return super().__repr__()
+
+    def top(self):
+        self.lower = -inf
+        self.upper = inf
+        return self
+
+    def is_top(self) -> bool:
+        return self._lower == -inf and self._upper == inf
+
+    def _less_equal(self, other: 'IntervalLattice') -> bool:
+        return other.lower <= self.lower and self.upper <= other.upper
+
+    def _meet(self, other: 'IntervalLattice'):
+        self.lower = max(self.lower, other.lower)
+        self.upper = min(self.upper, other.upper)
+        if self.lower > self.upper:
+            self.bottom()
+        return self
+
+    def _join(self, other: 'IntervalLattice') -> 'IntervalLattice':
+        self.lower = min(self.lower, other.lower)
+        self.upper = max(self.upper, other.upper)
+        return self
+
+    def _widening(self, other: 'IntervalLattice'):
+        if other.lower < self.lower:
+            self.lower = inf
+        if other.upper > self.upper:
+            self.upper = inf
+        return self
+
+    @classmethod
+    def evaluate(cls, expr: Expression):
+        """Evaluates an expression without variables, interpreting constants in the interval domain.
+        
+        If this method encounters any variables, it raises a ``ValueError``."""
+        return cls._visitor.visit(expr)
+
+    class Visitor(ExpressionVisitor):
+        """A visitor to abstractly evaluate an expression (without variables) in the interval domain."""
+
+        # noinspection PyMethodOverriding
```

Comment:
because the code gets very unreadable, since then I have to 'parse' the arguments inside the method, a functionality which is much better handled by explicit parameter names